micrometer-core/src/main/java/io/micrometer/core/instrument/binder/kafka/AbstractKafkaMetrics.java

```diff
@@ -0,0 +1,105 @@
+/**
+ * Copyright 2019 Pivotal Software, Inc.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package io.micrometer.core.instrument.binder.kafka;
+
+import io.micrometer.core.instrument.Gauge;
+import io.micrometer.core.instrument.Meter;
+import io.micrometer.core.instrument.Metrics;
+import io.micrometer.core.instrument.Tag;
+
+import org.apache.kafka.common.MetricName;
+import org.apache.kafka.common.metrics.KafkaMetric;
+import org.apache.kafka.common.metrics.MetricsReporter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.invoke.MethodHandles;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static java.util.stream.Collectors.toSet;
+
+/**
+ * @author Oleksii Bondar
+ */
+public abstract class AbstractKafkaMetrics implements MetricsReporter {
+
+    private final Logger logger = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+    private static final String METRIC_GROUP_NAME = "metric-group";
+
+    private static final Object STUB_VALUE = new Object();
+
+    // used to keep track of already registered/observable metrics to prevent registering existing metrics
+    private Map<String, Object> observableMetrics = new ConcurrentHashMap<>();
+
+    public abstract String getMetricPrefix();
+
+    public void configure(Map<String, ?> configs) {
+        // nothing needed
+    }
+
+    public void init(List<KafkaMetric> metrics) {
+        if (metrics != null) {
+            for (KafkaMetric kafkaMetric : metrics) {
+                metricChange(kafkaMetric);
+            }
+        }
+    }
+
+    public void metricChange(KafkaMetric metric) {
+        MetricName metricNameRef = metric.metricName();
+        String metricName = prepareMetricName(metricNameRef.name());
+        if (!observableMetrics.containsKey(metricName) && metric.metricValue() instanceof Double) {
```

Comment:
I don't see how this would work.

this is called with metricChange(global metric), and then afterwards with per topic metrics. You'll be getting the global ones in some cases and the non global ones in other cases? Depending on which is called first (which is arguably most likely the global ones, still..)

I'd suspect the more interesting ones are the per topic metrics.

The tag's for per topic metrics and global metrics are also not the same. So it'd require some magic to deal with node-id only existing in half of the numbers reported here for example.

Reply:
This is indeed a problem. Some metrics won't be registered with the current implementation, and if all were registered, at least the Prometheus registry would throw an error because Prometheus does not allow different sets of tag keys for the same metric name.

Reply:
Why not define `observableMetrics` as `Set<MetricName>` ?

Reply:
Is this PR still hanging on this?
Is it an option to define 2 separate binders, one for _global_ and one for _topic_? Users can bind them both/individually through `metric.reporters`.

I have a feeling this issue should ideally be solved at a higher level than this. Surely, this pattern (different tags on different occasions) is not Kafka-specific?